src/api/ml/CMakeLists.txt

```diff
@@ -0,0 +1,320 @@
+find_package(OCaml REQUIRED)
+
+set(exe_ext ${CMAKE_EXECUTABLE_SUFFIX})
+set(so_ext ${CMAKE_SHARED_LIBRARY_SUFFIX})
+set(bc_ext ".byte")
+
+set(z3ml_src ${CMAKE_CURRENT_SOURCE_DIR})
+set(z3ml_bin ${CMAKE_CURRENT_BINARY_DIR})
+
+if (Z3_BUILD_OCAML_EXTERNAL_LIBZ3)
+  add_custom_target(libz3_z3ml
+    ALL
+    DEPENDS ${Z3_BUILD_OCAML_EXTERNAL_LIBZ3}/libz3${so_ext}
+  )
+  set(libz3_path ${Z3_BUILD_OCAML_EXTERNAL_LIBZ3})
+else()
+  add_custom_target(libz3_z3ml
+    ALL
+    DEPENDS libz3
+  )
+  set(libz3_path ${PROJECT_BINARY_DIR})
+endif()
+
+add_custom_command(
+  OUTPUT
+    ${z3ml_bin}/z3native.ml
+    ${z3ml_bin}/z3native_stubs.c
+  COMMAND "${Python3_EXECUTABLE}"
+    "${PROJECT_SOURCE_DIR}/scripts/update_api.py"
+    ${Z3_FULL_PATH_API_HEADER_FILES_TO_SCAN}
+    "--ml-src-dir"
+    "${CMAKE_CURRENT_SOURCE_DIR}"
+    "--ml-output-dir"
+    "${CMAKE_CURRENT_BINARY_DIR}"
+  DEPENDS
+    ${PROJECT_SOURCE_DIR}/scripts/update_api.py
+    ${Z3_FULL_PATH_API_HEADER_FILES_TO_SCAN}
+    ${Z3_GENERATED_FILE_EXTRA_DEPENDENCIES}
+  COMMENT "Generatinging ${z3ml_bin}/z3native.ml and ${z3ml_bin}/z3native_stubs.c"
+  VERBATIM
+)
+
+add_custom_command(
+  OUTPUT 
+    ${z3ml_bin}/z3enums.ml
+  COMMAND "${Python3_EXECUTABLE}"
+    "${PROJECT_SOURCE_DIR}/scripts/mk_consts_files.py"
+    ${Z3_FULL_PATH_API_HEADER_FILES_TO_SCAN}
+    "--ml-output-dir"
+    "${CMAKE_CURRENT_BINARY_DIR}"
+  DEPENDS
+    ${PROJECT_SOURCE_DIR}/scripts/mk_consts_files.py
+    ${Z3_FULL_PATH_API_HEADER_FILES_TO_SCAN}
+    ${Z3_GENERATED_FILE_EXTRA_DEPENDENCIES}
+  COMMENT "Generating ${z3ml_bin}/z3enums.ml"
+  VERBATIM
+)
+
+add_custom_command(
```

Comment:
It's subtle and I prefer to copy them. The reason is `ocamlc` may generate some intermediate files at the source location. It general, it can also simplify the dependencies of custom commands when some files are generated to be in the build path while some files are at the source path.

Reply:
Oh yes, that's right. We do have to make sure everything is generated in the build directory, and sometimes that might indeed require copying source files around if the OCaml tools don't provide commandline flags to set the output path.